### ISSUE TITLE
streamline use of 'select new' with record types

### ISSUE THREAD

**Issue by gavinking** [jakartaee/persistence#420]
I would like to propose that we let people write:

```java
record BookSummary(String title, String author) {}
em.createQuery("select book.title, book.author.name from Book book", BookSummary.class).getResultList();
```

instead of forcing them into the extra ceremony of:

```java
record BookSummary(String title, String author) {}
em.createQuery("select new BookSummary(book.title, book.author.name) from Book book", BookSummary.class).getResultList();
```

This isn't critical, of course, but it's a pretty easy simplification to the user experience.

**Comment by gavinking** [comment]
Yes, sure, it would also be allowed for non-records. Sorry for being unclear.

**Comment by trajano** [comment]
Nothing to be sorry about @gavinking  I presume the purpose of this repo is to discuss things to make sure things are clear and less *behaviour is unspecified and vendor dependent*

**Comment by gavinking** [comment]
Another thing to consider as part of this is to just do certain sorts of implicit type conversions, for example, allow:

```java
int count = em.createSelectionQuery("select count(this) from Book", int.class).getSingleResult();
```

which would currently throw because `count(this)` is of type `Long`.

**Comment by njr-11** [comment]
This would be nice to add to Jakarta EE 12 where Jakarta Data is adding the ability to select multiple entity attributes and retrieve them as a Java record.  With this feature from Jakarta Persistence, a Jakarta Persistence-backed Jakarta Data provider would not need to edit the SELECT clause to insert the `NEW org.example.MyRecord(` part and could just use the original query without modifying it.

**Comment by gavinking** [comment]
@njr-11 Agreed.

**Comment by gavinking** [comment]
I had almost forgotten about this one, but it's important for Jakarta Data. Need to start work on it.

**Comment by gavinking** [comment]
I finally got around to writing this up.

> Also need to think about allowing this with `createNativeQuery()`.

I'm thinking of leaving that out for now.

**Comment by gavinking** [comment]
This was done by #824.
